File: app/src/main/java/com/asmat/rolando/popularmovies/database/DatabaseManager.java
package com.asmat.rolando.popularmovies.database;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.db.SupportSQLiteDatabase;
import android.arch.persistence.room.Room;
import android.arch.persistence.room.migration.Migration;
import android.content.Context;
import android.os.AsyncTask;

public enum DatabaseManager {

    INSTANCE;

    private AppDatabase db;
    private final String DATABASE_NAME = "movies-database";

    public AppDatabase getInstance() {
        return this.db;
    }

    public void setInstance(Context applicationContext) {
        AppDatabase db = Room.databaseBuilder(applicationContext,
                AppDatabase.class, DATABASE_NAME)
                .addMigrations(MIGRATION_1_2)
                .build();
        this.db = db;
    }

    /**
     * Favorite movies
     */

    public LiveData<FavoriteMovie> getFavoriteMovie(int id) {
        MoviesDAO dao = db.moviesDAO();
        LiveData<FavoriteMovie> favoriteMovie = dao.findFavoriteMovie(id);
        return favoriteMovie;
    }

    public void deleteFavoriteMovie(final FavoriteMovie movie) {
        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... voids) {
                MoviesDAO dao = db.moviesDAO();
                dao.deleteFavoriteMovie(movie);
                return null;
            }
        }.execute();
    }

    public void addFavoriteMovie(final FavoriteMovie favoriteMovie) {
        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... voids) {
                MoviesDAO dao = db.moviesDAO();
                dao.insertFavoriteMovie(favoriteMovie);
                return null;
            }
        }.execute();
    }

    public LiveData<Movie[]> getFavoriteMovies() {
        MoviesDAO dao = db.moviesDAO();
        return dao.loadAllFavoriteMovies();
    }

    /**
     * Watch Later movies
     */

    public LiveData<WatchLaterMovie> getWatchLaterMovie(int id) {
        MoviesDAO dao = db.moviesDAO();
        LiveData<WatchLaterMovie> watchLaterMovie = dao.findWatchLaterMovie(id);
        return watchLaterMovie;
    }

    public void deleteWatchLaterMovie(final WatchLaterMovie movie) {
        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... voids) {
                MoviesDAO dao = db.moviesDAO();
                dao.deleteWatchLaterMovie(movie);
                return null;
            }
        }.execute();
    }

    public void addWatchLaterMovie(final WatchLaterMovie watchLaterMovie) {
        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... voids) {
                MoviesDAO dao = db.moviesDAO();
                dao.insertWatchLaterMovie(watchLaterMovie);
                return null;
            }
        }.execute();
    }

    public LiveData<Movie[]> getWatchLaterMovies() {
        MoviesDAO dao = db.moviesDAO();
        return dao.loadAllWatchLaterMovies();
    }

    /**
     * Movie
     */
    public void insertMovie(final Movie movie) {
        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... voids) {
                MoviesDAO dao = db.moviesDAO();
                dao.insertMovie(movie);
                return null;
            }
        }.execute();
    }

    public void deleteMovie(final Movie movie) {
        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... voids) {
                MoviesDAO dao = db.moviesDAO();
                dao.deleteMovie(movie);
                return null;
            }
        }.execute();
    }

    /**
     * Migrations
     */

    static final Migration MIGRATION_1_2 = new Migration(1, 2) {
        @Override
        public void migrate(SupportSQLiteDatabase database) {
            database.execSQL(
                    "CREATE TABLE `watch_later_movies` (`id` INTEGER NOT NULL, "
                    + " PRIMARY KEY(`id`),"
                    + " FOREIGN KEY(`id`) REFERENCES `movies`(`id`) ON UPDATE NO ACTION ON DELETE CASCADE)");
        }
    };
}

File: app/src/main/java/com/asmat/rolando/popularmovies/database/DatabaseManager.kt
package com.asmat.rolando.popularmovies.database

import androidx.lifecycle.LiveData
import com.asmat.rolando.popularmovies.database.entities.FavoriteMovie
import com.asmat.rolando.popularmovies.database.entities.WatchLaterMovie
import io.reactivex.Completable
import io.reactivex.Scheduler
import io.reactivex.Single

/**
 * Database CRUD operations
 */
open class DatabaseManager(private val dao: MoviesDAO,
                           private val backgroundScheduler: Scheduler) {

    /**
     * Favorite movies
     */

    open fun addFavoriteMovie(favoriteMovie: FavoriteMovie): Completable {
        return dao.insertFavoriteMovie(favoriteMovie).subscribeOn(backgroundScheduler)
    }

    open fun getFavoriteMovie(id: Int): LiveData<FavoriteMovie> {
        return dao.findFavoriteMovie(id)
    }

    open fun deleteFavoriteMovie(id: Int): Single<Int> {
        return dao.deleteFavoriteMovie(id).subscribeOn(backgroundScheduler)
    }

    open fun getAllFavoriteMovies(): LiveData<List<FavoriteMovie>> {
        return dao.loadAllFavoriteMovies()
    }

    /**
     * Watch Later movies
     */

    open fun addWatchLaterMovie(watchLaterMovie: WatchLaterMovie): Completable {
        return dao.insertWatchLaterMovie(watchLaterMovie).subscribeOn(backgroundScheduler)
    }

    open fun getWatchLaterMovie(id: Int): LiveData<WatchLaterMovie> {
        return dao.findWatchLaterMovie(id)
    }

    open fun deleteWatchLaterMovie(id: Int): Single<Int> {
        return dao.deleteWatchLaterMovie(id).subscribeOn(backgroundScheduler)
    }

    open fun getAllWatchLaterMovies(): LiveData<List<WatchLaterMovie>> {
        return dao.loadAllWatchLaterMovies()
    }

}